### PR TITLE
#468 wire up create wp form

### PR DESCRIPTION
**Pull request by jpwaves**
Need review, will clean up chicken scratch code when I get back from work. Still need to fix `work-package-create` tests to conform to the new schema that uses WBS Numbers instead of project ID.

### CODE REVIEW

**Comment by jpwaves** [comment]
will fix merge conflicts when i get back from co-op as well

File: src/components/projects/create-wp-form/create-wp-form/create-wp-form.tsx [hunk 1]
```diff
@@ -41,24 +44,57 @@ const CreateWPFormView: React.FC<CreateWPFormViewProps> = ({
...
                     <Form.Label htmlFor='duration'>Duration</Form.Label>
                     <InputGroup>
-                      <Form.Control id='duration' name='duration' aria-label={'duration'} type='number' min={0} required></Form.Control>
+                      <Form.Control
+                        id='duration'
+                        name='duration'
+                        aria-label={'duration'}
+                        type='number'
+                        min={0}
+                        onChange={(e: any) => duration.setter(e.target.value)}
```
Comment (acd124) [inline src/components/projects/create-wp-form/create-wp-form/create-wp-form.tsx:96]: For all of these I believe the `e: any` isn't necessary since react provides the types, but for the number inputs you might want to parse the values so it's not setting them as strings. I'm not sure about this though and I haven't tested it.

**Review by acd124 — CHANGES_REQUESTED** [review]

File: src/components/projects/create-wp-form/create-wp-form/create-wp-form.tsx (hunk 1 above)
Comment (jpwaves) [reply src/components/projects/create-wp-form/create-wp-form/create-wp-form.tsx:96]: Right now, my `handleSubmit` expects all the data to come in as strings, and it seems to work so far. I know we have plans to switch to using react forms in the future, so I'm just gonna leave it alone since it works for now and keep the number inputs as strings.

File: src/components/projects/create-wp-form/create-wp-form.tsx [hunk 2]
```diff
@@ -75,20 +98,88 @@ const CreateWPForm: React.FC = () => {
...
+      }
+      const depWbsNums = dependencies.map(dependency => {
+        const depWbsNum = validateWBS(dependency.trim());
+        return {
+          carNumber: depWbsNum.car,
+          projectNumber: depWbsNum.project,
+          workPackageNumber: depWbsNum.workPackage
+        }
+      });
+      if (depWbsNums) {
```
Comment (acd124) [inline src/components/projects/create-wp-form/create-wp-form.tsx:138]: The typings say that `depWbsNums` will always result in an array, so it will always be truthy and the else won't be reached, not sure what you're testing here

File: src/utils/src/types/api-utils-types.ts [hunk 3]
```diff
@@ -42,21 +42,21 @@ export const enumType = <Items extends Readonly<string[]>>(...items: Items) =>
...
+      type: 'object',
+      properties: {
+        'user-agent': stringType
+        // Cookie: stringType
+      },
+      required: ['user-agent']
+    }
+  },
+  required: ['body', 'headers']
+} as const);
```
Comment (acd124) [inline src/utils/src/types/api-utils-types.ts:59]: I think the formatting here will go back if you save it again with prettier, mind if you check some of the files you made changes to?

**Review by acd124 — CHANGES_REQUESTED** [review]

File: src/backend/functions/work-packages-create.ts [hunk 4]
```diff
@@ -31,32 +31,67 @@ export const createWorkPackage: Handler<FromSchema<typeof inputSchema>> = async
...
+            carNumber: ele.carNumber,
+            projectNumber: ele.projectNumber,
+            workPackageNumber: ele.workPackageNumber
+          }
+        }
+      });
+    })
+  );
+
+  if (dependenciesWBSElems === null) throw new TypeError('One of the dependencies was not found.');
```
[the remark this answers is not in the document]
Comment (jpwaves) [reply src/backend/functions/work-packages-create.ts:82]: Oh gotcha, that should be an easy fix

File: src/utils/src/types/api-utils-types.ts (hunk 3 above)
Comment (jpwaves) [reply src/utils/src/types/api-utils-types.ts:59]: Yeah, I can go check. I think there's also a command that does all the prettier fixes for you (`prettier ./src --fix` I think? Or something like that) which we should consider adding to the `package.json` scripts so that it's easier to fix these prettier issues in the future. I think there's also one for ESLint, but I'd have to check

File: src/components/projects/create-wp-form/create-wp-form.tsx [hunk 5]
```diff
@@ -75,20 +98,88 @@ const CreateWPForm: React.FC = () => {
...
     // exits handleSubmit if form input invalid (should be changed in wire up)
-    const wbsNum: WbsNumber = validateWBS(e.target.wbsNum.value.trim());
+    let wbsNum: WbsNumber;
+    try {
+      wbsNum = validateWBS(projectWbsNum);
 
-    if (!isProject(wbsNum!)) {
-      alert('Please enter a valid Project WBS Number.');
-      return;
+      const { userId } = authContext!.user!;
```
[the remark this answers is not in the document]
Comment (jpwaves) [reply src/components/projects/create-wp-form/create-wp-form.tsx:124]: Found the hook, I'll add it in

File: src/utils/src/types/api-utils-types.ts (hunk 3 above)
Comment (acd124) [reply src/utils/src/types/api-utils-types.ts:59]: Yeah, great idea, I submitted a ticket for that

**Review by acd124 — COMMENTED** [review]
I don't think I have any more issues with the code that I'd like to bring up. I know this does affect the front end, so I'm not sure if we can actually see the changes in action, or if that's a not yet for this ticket. Other than that I think it's just the `console.log`s that would get removed unless there are more changes to be made

File: src/backend/functions/work-packages-create.ts [hunk 6]
```diff
@@ -31,32 +31,66 @@ export const createWorkPackage: Handler<FromSchema<typeof inputSchema>> = async
...
+        carNumber,
+        projectNumber,
+        workPackageNumber
+      }
+    }
+  });
+
+  if (wbsElem === null) throw new TypeError('No corresponding WBS Element for WBS Number.');
+
   const project = await prisma.project.findUnique({
```
Comment (jamescd18) [inline src/backend/functions/work-packages-create.ts:34]: I think this query is unneeded because we can just add something like "include: project: true" to the lookup via WBS number. Then we'd just want to add a check before that to ensure that the supplied work package number is 0 to ensure that resulting project field on the WBS element will not be null

File: src/backend/tests/work-packages-create.test.ts [hunk 7]
```diff
@@ -13,14 +13,20 @@ describe('work package create', () => {
     };
 
     describe('validate inputs', () => {
+      const mockWbsNum = {
```
Comment (jamescd18) [inline src/backend/tests/work-packages-create.test.ts:16]: I'm pretty sure there's a few of these in the test data you can use

Comment (jamescd18) [reply src/backend/tests/work-packages-create.test.ts:16]: https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/main/src/test-support/test-data/wbs-numbers.stub.ts

File: src/backend/tests/work-packages-create.test.ts [hunk 8]
```diff
@@ -13,14 +13,20 @@ describe('work package create', () => {
...
       const goodBody = {
         userId: 1,
         name: 'name',
         crId: 2,
-        projectId: 1,
+        projectWbsNum: mockWbsNum,
         startDate: '2015-10-06',
         duration: 1,
-        wbsElementIds: [1],
+        dependencies: [mockWbsNum],
```
Comment (jamescd18) [inline src/backend/tests/work-packages-create.test.ts:29]: New separate ticket cuz this reminded me of additional edge case business logic to disallow https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/issues/535

File: src/components/projects/create-wp-form/create-wp-form.tsx [hunk 9]
```diff
@@ -3,23 +3,46 @@
...
 
 export interface EditableTextInputListUtils {
   add: (val: any) => void;
   remove: (idx: number) => void;
   update: (idx: number, val: any) => void;
 }
 
+export interface StateBundle {
+  state: any;
+  setter: (val: any) => void;
```
Comment (jamescd18) [inline src/components/projects/create-wp-form/create-wp-form.tsx:22]: Pretty sure there's a better definition for this type

Comment (jamescd18) [reply src/components/projects/create-wp-form/create-wp-form.tsx:22]: https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/main/src/components/change-requests/review-change-request/review-change-request/review-change-request.tsx#L15

Comment (jamescd18) [reply src/components/projects/create-wp-form/create-wp-form.tsx:22]: Also this interface belongs elsewhere probably. Maybe in the shared folder in a new file? Or something along the lines of component supports tools?

File: src/components/projects/create-wp-form/create-wp-form.tsx [hunk 10]
```diff
@@ -75,20 +98,83 @@ const CreateWPForm: React.FC = () => {
     }
   };
 
-  const handleSubmit = (e: any) => {
+  const stateBundleBuilder = (state: any, setter: any) => {
```
Comment (jamescd18) [inline src/components/projects/create-wp-form/create-wp-form.tsx:101]: Better types, also move to diff file

File: src/components/projects/create-wp-form/create-wp-form.tsx [hunk 11]
```diff
@@ -75,20 +98,83 @@ const CreateWPForm: React.FC = () => {
...
+          projectNumber: wbsNum.project,
+          workPackageNumber: wbsNum.workPackage
+        },
+        startDate,
+        duration,
+        dependencies: depWbsNums,
+        expectedActivities,
+        deliverables
+      });
+      history.push(`${routes.CHANGE_REQUESTS}`);
```
Comment (jamescd18) [inline src/components/projects/create-wp-form/create-wp-form.tsx:153]: If not modifying the route, drop the template string and just have the variable

File: src/components/projects/create-wp-form/create-wp-form/create-wp-form.test.tsx [hunk 12]
```diff
@@ -3,24 +3,36 @@
...
-const mockDeliverables = ["go vroom"];
+const mockDependencies = ['1.2.3', '1.2.1', '1.12.0'];
+const mockEA = ['yooooo', 'wassup', 'bruh', 'whoosh'];
+const mockDeliverables = ['go vroom'];
 const mockUtils = {
   add: () => null,
   remove: () => null,
   update: () => null
 };
+const mockStateBundle = {
```
Comment (jamescd18) [inline src/components/projects/create-wp-form/create-wp-form/create-wp-form.test.tsx:17]: Pretty sure these can go in the proper test support utils stub file https://github.com/Northeastern-Electric-Racing/PM-Dashboard-v2/blob/main/src/test-support/test-data/test-utils.stub.ts

**Review by jamescd18 — CHANGES_REQUESTED** [review]
This is looking really good! There's a few different changes I've requested, after which I think it'll ready for final review. Amazing job Justin!! This is definitely a tough one doing modifications to the back-end, but I think you've done a great job learning and adapting!

File: src/components/projects/create-wp-form/create-wp-form.tsx (hunk 9 above)
Comment (jpwaves) [reply src/components/projects/create-wp-form/create-wp-form.tsx:22]: Yeah, I realized after typing that out that I didn't actually need the state at all, so I'm going to remove that interface and just use the type from your referenced example.

File: src/components/projects/create-wp-form/create-wp-form.tsx (hunk 10 above)
Comment (jpwaves) [reply src/components/projects/create-wp-form/create-wp-form.tsx:101]: Function is no longer needed after removing the `StateBundle` interface

File: src/backend/functions/work-packages-create.ts [hunk 13]
```diff
@@ -31,32 +31,67 @@ export const createWorkPackage: Handler<FromSchema<typeof inputSchema>> = async
...
+        projectNumber,
+        workPackageNumber
+      }
+    },
     include: {
-      wbsElement: true,
-      workPackages: { include: { wbsElement: true, dependencies: true } }
+      project: {
+        include: {
+          wbsElement: true,
```
Comment (jamescd18) [inline src/backend/functions/work-packages-create.ts:49]: WbsElement is the parent of project, so by doing `include` within `project`, you’re saying get A > get B > get A and duplicating pulling the wbsElement

File: src/components/projects/create-wp-form/create-wp-form/create-wp-form.test.tsx [hunk 14]
```diff
@@ -3,16 +3,19 @@
...
-const mockDeliverables = ["go vroom"];
-const mockUtils = {
-  add: () => null,
-  remove: () => null,
-  update: () => null
+import { mockUtils } from '../../../../test-support/test-data/test-utils.stub';
+import { render, screen } from '../../../../test-support/test-utils';
+import CreateWPFormView from './create-wp-form';
+
+const mockDependencies = ['1.2.3', '1.2.1', '1.12.0'];
```
Comment (jamescd18) [inline src/components/projects/create-wp-form/create-wp-form/create-wp-form.test.tsx:10]: This could probably just pull from the wbsNumber stub file

**Review by jamescd18 — CHANGES_REQUESTED** [review]

File: src/backend/functions/work-packages-create.ts (hunk 13 above)
Comment (jpwaves) [reply src/backend/functions/work-packages-create.ts:49]: Ah, I see. I'll remove that from the include

File: src/components/projects/create-wp-form/create-wp-form.tsx [hunk 15]
```diff
@@ -75,20 +93,76 @@ const CreateWPForm: React.FC = () => {
...
+        startDate,
+        duration,
+        dependencies: depWbsNums,
+        expectedActivities,
+        deliverables
+      });
+      history.push(routes.CHANGE_REQUESTS);
+    } catch (e) {
+      console.log(e);
+      alert('wbs num not valid');
```
Comment (jamescd18) [inline src/components/projects/create-wp-form/create-wp-form.tsx:144]:
```suggestion
      alert('something went wrong');
```

Comment (jamescd18) [reply src/components/projects/create-wp-form/create-wp-form.tsx:144]: this is because this catches any errors from the back-end call and so it won't always be because of the wbs nums

File: src/components/projects/create-wp-form/create-wp-form/create-wp-form.tsx [hunk 16]
```diff
@@ -40,34 +43,72 @@ const CreateWPFormView: React.FC<CreateWPFormViewProps> = ({
...
+                      onChange={(e) => crId(parseInt(e.target.value))}
+                      required
+                    ></Form.Control>
                   </Form.Group>
                 </Row>
                 <Row>
                   <Form.Group as={Col}>
-                    <Form.Label htmlFor='start-date'>Start Date</Form.Label>
-                    <Form.Control id='start-date' name='startDate' aria-label={'start date input'} type='date' required></Form.Control>
+                    <Form.Label htmlFor="start-date">Start Date</Form.Label>
```
Comment (jamescd18) [inline src/components/projects/create-wp-form/create-wp-form/create-wp-form.tsx:81]:
```suggestion
                    <Form.Label htmlFor="start-date">Start Date (YYYY-MM-DD)</Form.Label>
```

Comment (jamescd18) [reply src/components/projects/create-wp-form/create-wp-form/create-wp-form.tsx:81]: this is because it will error with different formats that people might think of using

**Review by jamescd18 — COMMENTED** [review]
this looks soooooo gooooood!!! and it works!!!! aaaahhhhhhhh!!! yaaaayy!!!

**Review by jamescd18 — APPROVED** [review]

**Review by acd124 — APPROVED** [review]
LGTM 👍🏼
